dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -319,30 +321,37 @@ JarLinkageReport generateLinkageReport(Path jarPath, SymbolReferenceSet symbolRe
    * the access modifiers in the {@code javaClass}.
    *
    * @see <a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-ClassModifier">
-   *   JLS 8.1.1. Class Modifiers</a>
+   *     JLS 8.1.1. Class Modifiers</a>
    */
-  private static boolean isClassAccessibleFrom(JavaClass javaClass, String sourceClassName) {
-    if (javaClass.isPublic()) {
-      // If a class is public, then it is accessible from anywhere
-      return true;
-    }
-    // Top-level class can only be declared as public or package private while nested-class can be
-    // declared as private or protected as well.
-    // https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-ClassModifier
-    if (javaClass.isPrivate()) {
-      // Class reference within same file is allowed to access private class. However, such case
-      // is already filtered at errorsFromSymbolReferences.
-      return false;
-    }
-    if (javaClass.isProtected()) {
-      // If an inner class is protected, it is accessible from within the package.
-      // Note that there are other cases where JVM bypass protected access (JLS 6.6.2)
-      // https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6.2
-      return ClassDumper.classesInSamePackage(javaClass.getClassName(), sourceClassName);
-    }
+  private boolean isClassAccessibleFrom(JavaClass javaClass, String sourceClassName)
+      throws ClassNotFoundException {
+    String targetClassName = javaClass.getClassName();
+    boolean publicOrSamePackage =
+        javaClass.isPublic() || ClassDumper.classesInSamePackage(targetClassName, sourceClassName);
+    if (javaClass.isNested()) {
+      // Nested class can be declared as private or protected, in addition to public and package
+      // private. Protected is treated same as package private.
+      // https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-ClassModifier
+      if (javaClass.isPrivate() || !publicOrSamePackage) {
```

Comment:
I think splitting `publicOrSamePackage` would make this easier to understand. And the way I think about the logic is maybe simpler?

1. If it's private, return false.
2. If it's public or in the same package as the source, return true iff the enclosing class is accessible.
3. Otherwise (not public and in a different package), return false.

Reply:
Updated. Does this look good now?

Reply:
I think this can be simpler if you don't switch on `javaClass.isNested()` at the top. If you take that away, and treat enclosingClassName() returning null as meaning you're at the top level, then you're left with:

```
if private:
  return false
if public or same package:
  if has enclosing class:
    return isAccessibleFrom(enclosing class, source class)
  else:
    return true
return false
```

Reply:
Updated in that way.